### PR TITLE
[Feature] Support STAR as an OpenMM model

### LINKED ISSUE #148: [Feature] Add support for the STAR model
Hi,

I would like to start contributing to this project. As we don't have a specific roadmap, I came up with a feature that seems reasonable - to add support for the [STAR model](https://star.is.tue.mpg.de/). I think it can be a good first feature to get to know the project and should prove useful as STAR is an improvement SMPL and can be used as a "drop-in replacement" for it.

Please let me know what you think.

### PR DESCRIPTION
**Pull request by kristijanbartol**
Implementation (integration) of the STAR model into MMHuman3D (resolves #148).

### Summary

Copy and slightly adapt the [original PyTorch implementation of the STAR model](https://github.com/ahmedosman/STAR/tree/master/star/pytorch).

### Changes

Compared to the original source code, I added explicit argument lists with their types. The model path can be loaded by specifying the directory where the gendered models are or by directly providing the path to the model.

### Discussion

In contrast to SMPL model, at the moment we cannot simply import the STAR model and then potentially extend it, because the [original repository](https://github.com/ahmedosman/STAR) does not properly expose STAR class to be imported. Therefore, it seems that the best solution was to copy and adapt the code.

The original license is now explicitly added as it feels somewhat improper to just copy the code, slightly adapt and then put another license. If we decide that it is OK, I will remove the second license.

### Future Work (Body Models Refactoring)

I noticed that the extended SMPL classes abuse the original SMPL in several places, for example:

1. The override implementation of the forward() method returns a Python dictionary instead SMPLOutput or ModelOutput as it is done in the original class. This forces the default dictionary syntax, such as model['vertices'] instead of model.vertices. Also, it creates code redundancy that we see in several places where the outputs are recreated as dictionaries, such as [here](https://github.com/open-mmlab/mmhuman3d/blob/main/mmhuman3d/models/body_models/smpl.py#L130) and [here](https://github.com/open-mmlab/mmhuman3d/blob/main/mmhuman3d/models/body_models/smpl.py#L327). The proper solution is to use an (extended) data class definition such as [this](https://github.com/vchoutas/smplx/blob/5fa20519735cceda19afed0beeabd53caef711cd/smplx/utils.py#L58).
2. The argument list for extended classes can be simplified. Instead of explicitly providing arguments, even [repeating the arguments](https://github.com/open-mmlab/mmhuman3d/blob/main/mmhuman3d/models/body_models/smpl.py#L281) of the [original method](https://github.com/vchoutas/smplx/blob/5fa20519735cceda19afed0beeabd53caef711cd/smplx/body_models.py#L308), the arguments can be kept as (*args, **kwargs) such as in [SPIN](https://github.com/nkolot/SPIN/blob/master/models/smpl.py#L14) or least we should avoid repeating the arguments from the original methods.
3. The [keypoint (re)mapping logic](https://github.com/open-mmlab/mmhuman3d/blob/main/mmhuman3d/models/body_models/smpl.py#L118) should be moved from the body models implementations as the keypoints have nothing to do with the body models once they are extracted from the vertices.
4. The gendered models such as [this](https://github.com/open-mmlab/mmhuman3d/blob/main/mmhuman3d/models/body_models/smpl.py#L206) seem redundant as the original SMPL model already accepts the gender parameter.
5. Improve unit tests (at least for the code coverage).

We can discuss these and other improvements in a separate issue.

### CODE REVIEW

**Comment by caizhongang** [comment]
@kristijanbartol Thank you for the contribution! I will review asap.

> ### Discussion
> For now, it seems sufficient to only import the STAR model, without extending the class like it is done with SMPL (the extended SMPL allows joint regressor argument to be provided, while there are no (known?) custom joint regressors for STAR).
> 
> **Help needed**: Please help me with verifying the unit tests I wrote. How do I properly run the tests locally?

To run the test locally, what we normally do is adding a main function in the test file and run it.

> ### Future Work (Body Models Refactoring)
> I noticed that the extended SMPL classes abuse the original SMPL in several places, for example:
> 
> 1. The override implementation of the forward() method returns a Python dictionary instead SMPLOutput or ModelOutput as it is done in the original class. This forces the default dictionary syntax, such as model['vertices'] instead of model.vertices. Also, it creates code redundancy that we see in several places where the outputs are recreated as dictionaries, such as [here](https://github.com/open-mmlab/mmhuman3d/blob/main/mmhuman3d/models/body_models/smpl.py#L130) and [here](https://github.com/open-mmlab/mmhuman3d/blob/main/mmhuman3d/models/body_models/smpl.py#L327). The proper solution is to use an (extended) data class definition such as [this](https://github.com/vchoutas/smplx/blob/5fa20519735cceda19afed0beeabd53caef711cd/smplx/utils.py#L58).
> 2. The argument list for extended classes can be simplified. Instead of explicitly providing arguments, even [repeating the arguments](https://github.com/open-mmlab/mmhuman3d/blob/main/mmhuman3d/models/body_models/smpl.py#L281) of the [original method](https://github.com/vchoutas/smplx/blob/5fa20519735cceda19afed0beeabd53caef711cd/smplx/body_models.py#L308), the arguments can be kept as (*args, **kwargs) such as in [SPIN](https://github.com/nkolot/SPIN/blob/master/models/smpl.py#L14) or least we should avoid repeating the arguments from the original methods.
> 3. The [keypoint (re)mapping logic](https://github.com/open-mmlab/mmhuman3d/blob/main/mmhuman3d/models/body_models/smpl.py#L118) should be moved from the body models implementations as the keypoints have nothing to do with the body models once they are extracted from the vertices.
> 4. The gendered models such as [this](https://github.com/open-mmlab/mmhuman3d/blob/main/mmhuman3d/models/body_models/smpl.py#L206) seem redundant as the original SMPL model already accepts the gender parameter.
> 5. Improve unit tests (at least for the code coverage).
> 
> We can discuss these and other improvements in a separate issue.

Cool, I would suggest opening up a new discussion [here](https://github.com/open-mmlab/mmhuman3d/discussions).

**Comment by kristijanbartol** [comment]
> To run the test locally, what we normally do is adding a main function in the test file and run it.

I found it more convenient to create a separate main module in the root directory, but I guess the idea is the same.

The problem is that when I try to import mmhuman3d, I immediately get an error on [this line](https://github.com/open-mmlab/mmhuman3d/blob/main/mmhuman3d/__init__.py#L1). Obviously I don't have mmcv as it's [not listed as a dependency](https://github.com/open-mmlab/mmhuman3d/tree/main/requirements).

Can you please help me with this? I apologize as this is probably due to my inexperience with Python packages, but I don't understand how could mmcv be imported if not installed.

**Comment by caizhongang** [comment]
> > To run the test locally, what we normally do is adding a main function in the test file and run it.
> 
> I found it more convenient to create a separate main module in the root directory, but I guess the idea is the same.
> 
> The problem is that when I try to import mmhuman3d, I immediately get an error on [this line](https://github.com/open-mmlab/mmhuman3d/blob/main/mmhuman3d/__init__.py#L1). Obviously I don't have mmcv as it's [not listed as a dependency](https://github.com/open-mmlab/mmhuman3d/tree/main/requirements).
> 
> Can you please help me with this? I apologize as this is probably due to my inexperience with Python packages, but I don't understand how could mmcv be imported if not installed.

I would suggest following our [install](https://github.com/open-mmlab/mmhuman3d/blob/main/docs/install.md) guide to set up the environment, it includes installation of mmcv-full.

**Comment by kristijanbartol** [comment]
I apologize for stumbling upon the trivial issues, but I didn't figure out how to load the STAR model files for the build system to find them. Obviously, the SMPL models are somehow loaded, even though the `./data` folder does not exist in the repository. I guess that the SMPL models are downloaded from somewhere, but I didn't find the lines of code where this is done. Thanks in advance!

**Comment by kristijanbartol** [comment]
OK, finally found it in `.github/workflows/build.yml`:

`wget -O mmhuman3d.7z -q https://openmmlab-share.oss-cn-hangzhou.aliyuncs.com/mmhuman3d/mmhuman3d.7z`

Inside of this file there are all the necessary data for running the automatic build and tests.

As I don't have the permission, can some of the owners add STAR male, female, and neutral models to `data/body_models/star`. They should be named simply `male.npz`, `female.npz`, and `neutral.npz`.

Thank you!

**Comment by caizhongang** [comment]
> OK, finally found it in `.github/workflows/build.yml`:
> 
> `wget -O mmhuman3d.7z -q https://openmmlab-share.oss-cn-hangzhou.aliyuncs.com/mmhuman3d/mmhuman3d.7z`
> 
> Inside of this file there are all the necessary data for running the automatic build and tests.
> 
> As I don't have the permission, **can some of the owners add STAR male, female, and neutral models to `data/body_models/star`?** They should be named simply `male.npz`, `female.npz`, and `neutral.npz`.
> 
> Thank you!

Hi @kristijanbartol sorry for the late reply as we are quite busy recently for the upcoming release.

Sure thing, but mmhuman3d will be downloaded every time for the build, so we'd prefere to include only the neutral model for unit tests.

**Comment by caizhongang** [comment]
> As I don't have the permission, **can some of the owners add STAR male, female, and neutral models to `data/body_models/star`?** They should be named simply `male.npz`, `female.npz`, and `neutral.npz`.

The mmhuman3d.7z has been updated, and `STAR_NEUTRAL.npz` is placed in `data/body_models/star`. Please take a look @kristijanbartol.

**Comment by caizhongang** [comment]
Hi @kristijanbartol , v0.7.0 has been released last Friday, and v0.8.0 is scheduled to be released by end of May. I think STAR is going to be a good feature in v0.8.0, shall we aim for it?

**Comment by kristijanbartol** [comment]
@caizhongang that would be great! I honestly got stuck with `ImportError: libcudart.so.10.1: cannot open shared object file: No such file or directory` and still didn't figure out how to fix it and why does it happen, but I'm sure I will work my way out of it soon. Thank you for your support!

**Comment by kristijanbartol** [comment]
The builds were canceled after 360m. It seems that it got stuck on test_smpl.py. My guess is that I am responsible for that because I added  `if __name__ == '__main__':`, which for some reason prevents the test from being completed.

I will remove this part both from test_smpl.py and test_star.py and then we could try to build again. I apologize for the inconvenience.

**Comment by kristijanbartol** [comment]
Hi @caizhongang , the build failed again by getting stuck on `test_smpl.py`. Now I'm getting the feeling that I might not be responsible. There is still a possibility that the problem is with `mmhuman3d/models/body_models/__init__.py` where I try to import STAR, which somehow makes the system get stuck.

There might be two possibilities:
1. I do something wrong when importing STAR, which is related to how the decorator `BODY_MODELS.register_module`, that depends on mmcv.cnn.MODELS, works, but I can't guess what could be wrong (note, I just follow what is done in `smpl.py`).
2. There is something wrong with the GitHub build system. Last successful build was done 10 days, before the 0.7.0 version release and it might be that occasionally the system needs to be cleaned. I don't know if you had any problems with it before, but I found a motivation for this [here](https://github.com/actions/virtual-environments/issues/1918#issuecomment-717878289).

**Comment by caizhongang** [comment]
> Update: I still need an approval to run the build

So sorry for the delay! All checks have passed. Let me know if I can help in anyway.

**Comment by kristijanbartol** [comment]
Soon we'll hopefully complete the PR so I'll be able to run the workflows without approval. :) Now I have changed one file that might be causing the build failure, so let's try to run it again!

**Comment by caizhongang** [comment]
> Soon we'll hopefully complete the PR so I'll be able to run the workflows without approval. :)

@kristijanbartol Hi Kristijan , I've temporarily lifted the restriction to allow all users to run workflow (except for new Github users) to assist you in the debugging. However, as each workflow run incurs a cost, we'd appreciate that if you can run workflow only if you are confident. For example, the linting may be fixed locally before pushing a new commit. Thanks! and let us know if we can assist in any way.

**Comment by kristijanbartol** [comment]
@caizhongang , thank you very much! I apologize for the mistake. I temporarily changed my computer and forgot to (re)install everything up for local checks.

**Comment by kristijanbartol** [comment]
@caizhongang , I think you can review the changes now, finally. I prefer writing a few more tests for new utils/transforms.py functions, as they are only covered indirectly from the test_star.py tests, but I consider that optional at this point (I can do it at the beginning of the next week, before the release).

Also, after we finish the review, I will squash all these commits into one or few. Thank you!

**Comment by caizhongang** [comment]
> @caizhongang , I think you can review the changes now, finally. I prefer writing a few more tests for new utils/transforms.py functions, as they are only covered indirectly from the test_star.py tests, but I consider that optional at this point (I can do it at the beginning of the next week, before the release).

@kristijanbartol , thank you for the update! Given this PR is an important one and deserves careful review, do you mind if we delay it slightly to the next version (v0.9.0)?

> Also, after we finish the review, I will squash all these commits into one or few. Thank you!

The PR will be automatically squashed into one commit when merging, don't worry :)

**Comment by kristijanbartol** [comment]
@caizhongang sure, I don't mind waiting for another release. In the meantime, I will complete unit tests. Also, I can make the PR more complete by adding the converters between SMPL and STAR, and providing brief instructions on how to use the models in body_models/README.md (serving also as some sort of functional test for the new code - I guess functional tests are not used in OpenMMLab repositories?).

**Comment by caizhongang** [comment]
> @caizhongang sure, I don't mind waiting for another release. In the meantime, I will complete unit tests. Also, I can make the PR more complete by adding the converters between SMPL and STAR, and providing brief instructions on how to use the models in body_models/README.md (serving also as some sort of informal functional test for the new code - I guess functional tests are not used in OpenMMLab repositories?).

I think the conversion tool is good idea, but we can have a new PR just for that (great to include other [conversion tools](https://github.com/vchoutas/smplx/tree/master/transfer_model)).

I would say some tests can be regarded as functional tests (multiple modules used in a mock pipeline, [example](https://github.com/open-mmlab/mmhuman3d/blob/main/tests/test_smplify.py))? But unit tests we use are generally simple, whose another purpose is to catch unexpected behaviors of new changes.

**Comment by kristijanbartol** [comment]
Do you think it would be a good idea to create a PR in parallel with this one where we include the conversion tools? I think it's more advanced than this PR, but I guess now I got used to the procedure so I should move faster.

**Comment by kristijanbartol** [comment]
@caizhongang A reminder to review this pull request for the next release :)

File: mmhuman3d/utils/transforms.py
Comment (caizhongang) [inline mmhuman3d/utils/transforms.py:632]: Can `_quat2mat` be replaced by the `quat_to_rotmat` implementation (line 190)?

File: mmhuman3d/utils/transforms.py
Comment (caizhongang) [inline mmhuman3d/utils/transforms.py:658]: Can `rodrigues` be replaced by `aa_to_rotmat` (line 59) implementation?

File: mmhuman3d/utils/transforms.py
Comment (caizhongang) [inline mmhuman3d/utils/transforms.py:676]: Nice function, two suggestions:
1. `with_zeros` might be unclear. It may be better called something along the line of `make_homegeneous_rotmat`
2. Can we make it more general that it also accepts a single transformation matrix (i.e. shape is (3, 4), no batch dim)

File: mmhuman3d/utils/transforms.py
Comment (caizhongang) [inline mmhuman3d/utils/transforms.py:551]: Can we also keep `transform_to_camera_frame` as there might be situations that we don't have a batch dim.

File: tests/test_models/test_body_models/test_utils.py
Comment (caizhongang) [inline tests/test_models/test_body_models/test_utils.py:11]: Same as above, can we keep `test_transform_to_camera_frame`?

File: mmhuman3d/models/body_models/star.py
Comment (caizhongang) [inline mmhuman3d/models/body_models/star.py:10]: Let's move the additional license to this [file](https://github.com/open-mmlab/mmhuman3d/blob/main/docs/additional_licenses.md).

File: mmhuman3d/models/body_models/star.py
Comment (caizhongang) [inline mmhuman3d/models/body_models/star.py:212]: Is it possible to follow other body models and
1. make the output in the dictionary form
2. add the keypoint remapping

I have provide some discussion point [here](https://github.com/open-mmlab/mmhuman3d/discussions/162)。 Appreciate your feedback. ;)

**Review by caizhongang — COMMENTED** [review]
Hi @kristijanbartol , firstly sorry for the delayed review as it has been a bit overwhelming for us recently due to some new features we are building. 

I think the PR looks great. Besides the comments below, I'm wondering if it is possible to add an HMR config for STAR (i.e., train HMR to generate STAR) because we do not have a use case for STAR at the moment? But we can have a separate PR to address this, for sure.

**Comment by kristijanbartol** [comment]
Hi @caizhongang , thank you for the comments, they are quite on-point and at the first glance I think I totally agree with them. I'm coming from CVPR on 27th and might try to adress the comments for the final review the same day.

**Comment by caizhongang** [comment]
> Hi @caizhongang , thank you for the comments, they are quite on-point and at the first glance I think I totally agree with them. I'm coming back from CVPR on 27th and might try to adress the comments for the final review the same day.

Sounds great! Enjoy your stay in CVPR!

**Comment by caizhongang** [comment]
Hi @kristijanbartol , hope you had a blast in CVPR. As the end of June is around the corner, hopefully we can merge the PR in time!

**Comment by kristijanbartol** [comment]
I made an upgrade of my local Ubuntu and I guess something had to change wrt conda environment I had, because now PyTorch3D is causing trouble. I tried to recreate the environment several times even with different Python versions, but still I get errors when I try to verify the PyTorch3D installation.

I'm so tired of PyTorch3D and its installation in general, because it's always so painful to get the correct combination of versions of PyTorch, CUDA toolkit, and PyTorch3D. I don't have any more time to spend on this and I have to go on doing other things.

I made most of the changes, but need to run the code to make STAR consistent with SMPL in terms of their dictionaries and expected behavior of their values. As I am unable to install the environment for the Nth time, I am really frustrated...

P.S. I'll try to find some time before the end of the tomorrow...

**Comment by caizhongang** [comment]
> I made an upgrade of my local Ubuntu and I guess something had to change wrt conda environment I had, because now PyTorch3D is causing trouble. I tried to recreate the environment several times even with different Python versions, but still I get errors when I try to verify the PyTorch3D installation. The other guess is that the latest conda package for PyTorch3D has been updated and it causes issues, but that doesn't explain why the original environement was problematic.

> I'm so tired of PyTorch3D and its installation in general, because it's always so painful to get the correct combination of versions of PyTorch, CUDA toolkit, and PyTorch3D. I don't have any more time to spend on this and I have to go on doing other things.

> I made most of the changes, but need to run the code to make STAR consistent with SMPL in terms of their dictionaries and expected behavior of their values. As I am unable to install the environment for the Nth time, I am really frustrated...

PyTorch3D indeed complicates the installation a lot, we are thinking of getting around it too (likely a complete overhaul of the rendering pipeline). For this PR, I guess you may remove all PyTorch3D related stuff locally (just don't commit these) and test STAR?

> P.S. I'll try to find some time before the end of tomorrow...

No rush, v0.9.0 release is delayed to 7 July. Please let us know if you still can't get PyTorch3D to install properly. Thanks!

**Comment by kristijanbartol** [comment]
You are absolutely right, the contributions should be fun! I pushed my changes and need to add a few more unit tests. Also, I will revisit all the changes tomorrow to make sure everything works as expected. I still wasn't able to install pytorch3d, though, so I will also describe my setup in one of the following comments... If you think you might not have time later, you might do another review now. If you'll have some time, then just wait a bit for me to verify the changes once again. Thank you for the support, @caizhongang

**Comment by kristijanbartol** [comment]
I don't understand the code coverage failure I'm getting. When I click on the report, for example, in [this file](https://app.codecov.io/gh/open-mmlab/mmhuman3d/compare/152/diff?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=checks&utm_campaign=pr%20comments&utm_term=open-mmlab#D1L559), the red and green lines don't make much sense and do not correspond with the total change in percentage. I also see the files that I didn't change and I believe have nothing to do with my changes. Can you please give me some advice regarding this?

P.S. Now I wrote all the planned unit tests for STAR, so the PR will be ready as soon as I solve the code coverage error.

**Comment by caizhongang** [comment]
> I don't understand the code coverage failure I'm getting. When I click on the report, for example, in [this file](https://app.codecov.io/gh/open-mmlab/mmhuman3d/compare/152/diff?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=checks&utm_campaign=pr%20comments&utm_term=open-mmlab#D1L559), the red and green lines don't make much sense and do not correspond with the total change in percentage. I also see the files that I didn't change and I believe have nothing to do with my changes. Can you please give me some advice regarding this?

@kristijanbartol Don't worry, it is acceptable if the patch coverage is above 85%. The PR looks good, I'll have a detailed review today and hopefully we can merge it. Thanks a lot for your contribution!

File: mmhuman3d/utils/transforms.py
Comment (caizhongang) [inline mmhuman3d/utils/transforms.py:551]: We may consider putting `transform_to_camera_frame` and `batch_transform_to_camera_frame` back to `mmhuman3d/models/body_models/utils.py` for two reasons: 

1. the `transforms.py` contains very general rotation convertions (e.g. axis-angle to rotmat), but  `transform_to_camera_frame` and `batch_transform_to_camera_frame` are very specific to body models.
2. some of our downstream projects are using `transform_to_camera_frame` and `batch_transform_to_camera_frame`, changing their locations may cause incompatibility

What do you think?

**Review by caizhongang — COMMENTED** [review]
The PR is almost ready! Just one last comment and we are good to go ;)

File: mmhuman3d/utils/transforms.py
Comment (kristijanbartol) [reply mmhuman3d/utils/transforms.py:551]: > the transforms.py contains very general rotation convertions (e.g. axis-angle to rotmat), but transform_to_camera_frame and batch_transform_to_camera_frame are very specific to body models.

I was considering to move it back, but yes, these transforms are very specific to body models.

> some of our downstream projects are using transform_to_camera_frame and batch_transform_to_camera_frame, changing their locations may cause incompatibility

This is even stronger reason. ✅

**Comment by kristijanbartol** [comment]
It's ready now!

**Review by caizhongang — APPROVED** [review]

**Pull request merged**
